sharkfuser/include/fusili/asm_emitter.h

```diff
@@ -0,0 +1,420 @@
+                                      const std::string &suffix) {
+  std::ostringstream oss;
+  std::vector<std::string> ssaValueNames;
+
+  // Emit `torch.constant.int` ops for each int value
+  for (size_t i = 0; i < listOfInts.size(); ++i) {
+    std::string ssaValueName =
+        "%" + prefix + "_val_" + std::to_string(i) + "_" + suffix;
+    oss << ssaValueName << " = torch.constant.int " << listOfInts[i]
+        << "\n    ";
+    ssaValueNames.push_back(ssaValueName);
+  }
+
+  // Emit the ListConstruct op
+  oss << "%" + prefix + "_" + suffix << " = torch.prim.ListConstruct ";
+  for (size_t i = 0; i < ssaValueNames.size(); ++i) {
+    if (i > 0)
+      oss << ", ";
+    oss << ssaValueNames[i];
+  }
+  oss << " : (";
+  for (size_t i = 0; i < ssaValueNames.size(); ++i) {
+    if (i > 0)
+      oss << ", ";
+    oss << "!torch.int";
+  }
+  oss << ") -> !torch.list<int>\n";
+
+  return oss.str();
+}
+
+//===----------------------------------------------------------------------===//
+//
+// TensorAttr ASM Emitter Methods
+//
+//===----------------------------------------------------------------------===//
+
+// Emits a ranked tensor type in MLIR assembly representation.
+//
+// This expects ranked tensors (non-scalar) as we blanket generate
+// a `!torch.vtensor` type. The caller is responsible to check for
+// this. In the future we may want to extend this (or add new methods)
+// for scalar types (such as `!torch.int` or `!torch.bool`).
+//
+// Example:
+//
+//    TensorAttr t;
+//    t.setName("tensor")
+//      .setDataType(DataType::Float)
+//      .setDim({2, 3})
+//      .setStride({3, 1})
+//
+//    t.getRankedTensorTypeAsm() generates "!torch.vtensor<[2,3],f32>"
+//
+inline std::string TensorAttr::getRankedTensorTypeAsm() const {
+  assert(!isScalar() &&
+         "TensorAttr::getRankedTensorTypeAsm expects a ranked tensor");
+  assert(!getDim().empty() &&
+         "TensorAttr::getRankedTensorTypeAsm expects non-empty dims");
+  assert(getDataType() != DataType::NotSet &&
+         "TensorAttr::getRankedTensorTypeAsm expects a valid data type");
+
+  std::ostringstream oss;
+  oss << "!torch.vtensor<[";
+  const std::vector<int64_t> &dims = getDim();
+  for (size_t i = 0; i < dims.size(); ++i) {
+    if (i > 0)
+      oss << ",";
+    oss << dims[i];
+  }
+  oss << "],";
+  oss << DataTypeToMlirTypeAsm.at(getDataType());
+  oss << ">";
+  return oss.str();
+}
+
+// Emits an MLIR SSA value name starting with the `%` sigil based off the
+// TensorAttr name but only using alphanumeric / underscore [A-Za-z0-9_]
+// characters.
+//
+// `foo_Bar::X0` would become `%foo_BarX0`
+inline std::string TensorAttr::getMlirSSAValueNameAsm() const {
+  assert(!getName().empty() &&
+         "TensorAttr name must not be empty for `getMlirSSAValueNameAsm`");
+
+  std::string filtered;
+  for (char c : getName()) {
+    if (std::isalnum(static_cast<unsigned char>(c)) || c == '_')
+      filtered += c;
+  }
```

Comment:
```suggestion
  std::string filtered;
  filtered.erase(remove_if(filtered.begin(), filtered.end(), [](char c) {
    return !isalnum(c) && c != '_';
  }), filtered.end());
```
If you feel like channeling Sean Parent 🤷‍♂️

Reply:
I had to squint at this, but isn't `filtered` an empty string to begin with, so what are we erasing?

Did you mean something like this?
```cpp
std::string filtered = getName();
filtered.erase(
    std::remove_if(filtered.begin(), filtered.end(),
        [](char c) {
            return !(std::isalnum(c) || c == '_');
        }
    ),
    filtered.end()
);
```

I didn't worry much about efficiency (wrt string reallocations) since we're only dealing with relatively small names here so went with code that was easy to follow, but happy to adopt this here.